Core/Core/Planner/Model/UseCase/GetPlannables+ObserverEvents.swift

```diff
@@ -0,0 +1,210 @@
+                        if case .teacher = request.env.app {
+                            let plannerNotes = self.plannerNotesPublisher(env: request.env, contexts: contexts)
+                            allPublisher = calendarEvents.merge(with: calendarAssignments, plannerNotes).eraseToAnyPublisher()
+                        } else {
+                            allPublisher = calendarEvents.merge(with: calendarAssignments).eraseToAnyPublisher()
+                        }
+
+                        return allPublisher
+                            .reduce([], { $0 + $1 })
+                            .map { events in
+                                return (request, EventsResponse(events: events, response: nil, error: nil))
+                            }
+                            .catch { failure in
+                                return Just((
+                                    request,
+                                    EventsResponse(events: nil, response: failure.response, error: failure.error)
+                                ))
+                            }
+                    }
+                    .eraseToAnyPublisher()
+            }
+            .sink { (request: EventsRequest, response: EventsResponse) in
+                request.callback(response.events, response.response, response.error)
+            }
+            .store(in: &subscriptions)
+    }
+
+    func getObserverEvents(env: AppEnvironment, callback: @escaping EventsRequest.Callback) {
+        observerEvents.send(EventsRequest(env: env, callback: callback))
+    }
+}
+
+// MARK: - Publishers
+
+extension GetPlannables {
+
+    private func contextsPublisher(env: AppEnvironment) -> AnyPublisher<[Context], Never> {
+        if let contexts = contextCodes?.compactMap({ Context(canvasContextID: $0) }) {
+            return Just(contexts).eraseToAnyPublisher()
+        }
+
+        guard let userID = userID else {
+            return Just([]).eraseToAnyPublisher()
+        }
+
+        let request = GetCoursesRequest(
+            enrollmentState: .active,
+            enrollmentType: .observer,
+            state: [.available],
+            perPage: 100
+        )
+
+        return Future<[Context], Never> { promise in
+
+            env.api.exhaust(request) { response, _, _ in
+                guard let courses = response else {
+                    promise(.success([]))
+                    return
+                }
+
+                var codes: [String] = []
+                for course in courses {
+                    let enrollments = course.enrollments ?? []
+                    for enrollment in enrollments where enrollment.associated_user_id?.value == userID {
+                        codes.append(Context(.course, id: course.id.value).canvasContextID)
+                    }
+                }
+
+                let contexts = codes.compactMap({ Context(canvasContextID: $0) })
```

Comment:
Would it be possible to replace `codes: [String]` with `contexts[Context]`, so we don't need to convert Context -> Code -> Context?

Reply:
Yes, definitely we can do that. This one slipped as I was moving code around. Thanks for noticing that. 👍